### ISSUE TITLE
ssh requires use of less-secure MAC

### ISSUE THREAD

**Issue by strowger** [blinksh/blink#129]
The "ssh hardening" text at https://stribika.github.io/2015/01/04/secure-secure-shell.html suggests that the following MACs are acceptably secure.

hmac-sha2-256-etm@openssh.com
hmac-ripemd160-etm@openssh.com
umac-128-etm@openssh.com,hmac-sha2-512
hmac-sha2-256,hmac-ripemd160
umac-128@openssh.com

Giving an /etc/ssh/sshd_config line:
MACs hmac-sha2-256-etm@openssh.com,hmac-ripemd160-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,hmac-ripemd160,umac-128@openssh.com

If this configuration is used, blink will fail to connect to the host, complaining
"SSHSession:DEBUG:Unable to exchange encryption keys".

You do support a "safe" selection of ciphers and KexAlgorithms, and I appreciate that this problem is upstream in your ssh library. You are already doing better than nearly all ssh clients for mobile devices.

**Comment by carloscabanero** [comment]
Thanks for adding this one! I might hire someone to implement Ed25519 and this. Any specific method?

**Comment by strowger** [comment]
sorry, i'm not sure what you mean by 'method'. 

if your upstream ssh library has no plans to add these then i guess paying someone might be good (assuming sales are good enough to make it worthwhile).

**Comment by agh** [comment]
@strowger I think you're seeing a Key Exchange issue, rather than Message Authentication. Message you're getting from the remote server complains about *exchange* and I commented on #60 which is similar to this issue with my sshd_config -

```
KexAlgorithms curve25519-sha256@libssh.org
Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,umac-128@openssh.com
```

I needed to add another option to KexAlgorithms in order to be able to connect.

I've noted which one works on #60 and also dropped in a quick script snipped which will regenerate your moduli to be secure, vs. using those distributed with your operating system. Hope this helps.
